File: management-frontend/src/utils/wholesaleOrderWorkflow.ts
import type { TFunction } from 'i18next';
import type { AuditLog, WholesaleOrder } from '../types';
import { allOrderLinesFullyAssigned, shipmentStatusAllowsAssignmentChange } from './wholesaleOrderAssignment';
import {
  getWholesaleOrderEmailAudits,
  isWholesaleOrderEmailSkippedAudit,
  isWholesaleOrderEmailSentAudit,
  isWholesaleOrderInvoiceEmailDone,
  orderHasInvoiceDocument,
  parseWholesaleOrderEmailAuditBase,
} from './wholesaleOrderEmail';
import { shipmentHasDeliveryNoteStarted } from './shipmentStatus';

export type WholesaleOrderProcessStepKey =
  | 'stepCreate'
  | 'stepOrderConfirmation'
  | 'stepStartShipment'
  | 'stepFinishShipment'
  | 'stepSendInvoiceEmail'
  | 'stepPaymentConfirmation'
  | 'stepComplete';

export type WholesaleOrderStatusChipColor =
  | 'success'
  | 'warning'
  | 'primary'
  | 'error'
  | 'default'
  | 'secondary'
  | 'info';

export interface WholesaleOrderProcessStep {
  labelKey: WholesaleOrderProcessStepKey;
  done: boolean;
}

export interface WholesaleOrderWorkflowContext {
  auditLogs?: AuditLog[];
  /** When set with totalProofAmount, payment step stays open until fully received. */
  orderTotal?: number;
  totalProofAmount?: number;
}

function shipmentMetrics(order: WholesaleOrder) {
  const shipments = order.shipments ?? [];
  const hasShipments = shipments.length > 0;
  const allShipmentsStarted =
    hasShipments && shipments.every((s) => shipmentHasDeliveryNoteStarted(s));
  const allShipmentsCompleted =
    hasShipments && shipments.every((s) => s.status === 'completed');
  return { hasShipments, allShipmentsStarted, allShipmentsCompleted };
}

function hasPaymentProofDocument(order: WholesaleOrder): boolean {
  return (
    !!order.payment_proof_url?.trim() ||
    (order.documents?.some((d) => d.type === 'payment_proof') ?? false)
  );
}

export function wholesaleOrderGrandTotal(order: WholesaleOrder): number {
  const itemsTotal =
    order.total_net ?? order.items?.reduce((sum, it) => sum + (it.line_total || 0), 0) ?? 0;
  return itemsTotal + (Number(order.shipping_fee) || 0);
}

function resolveTotalProofAmount(
  order: WholesaleOrder,
  ctx: WholesaleOrderWorkflowContext,
): number | undefined {
  if (ctx.totalProofAmount != null && Number.isFinite(ctx.totalProofAmount)) {
    return ctx.totalProofAmount;
  }
  const fromOrder = order.workflow_payment_proof_total;
  if (fromOrder != null && Number.isFinite(fromOrder)) {
    return fromOrder;
  }
  return undefined;
}

export function buildWholesaleOrderWorkflowContext(
  order: WholesaleOrder,
  auditLogs?: AuditLog[],
): WholesaleOrderWorkflowContext {
  const orderTotal = wholesaleOrderGrandTotal(order);
  if (auditLogs?.length) {
    return {
      auditLogs,
      orderTotal,
      totalProofAmount: computeTotalProofAmountFromAudits(order, auditLogs),
    };
  }
  const totalProofAmount = resolveTotalProofAmount(order, { orderTotal });
  return totalProofAmount != null ? { orderTotal, totalProofAmount } : { orderTotal };
}

function parseAuditChanges(raw: string): Record<string, unknown> | null {
  try {
    const parsed = JSON.parse(raw) as unknown;
    if (!parsed || typeof parsed !== 'object') return null;
    const obj = parsed as Record<string, unknown>;
    const nested = obj.changes;
    if (nested && typeof nested === 'object') {
      return nested as Record<string, unknown>;
    }
    return obj;
  } catch {
    return null;
  }
}

/** Mirrors detail-page payment proof total from upload audits. */
export function computeTotalProofAmountFromAudits(
  order: WholesaleOrder,
  auditLogs: AuditLog[],
): number {
  const proofDocs = order.documents?.filter((d) => d.type === 'payment_proof') ?? [];
  if (proofDocs.length === 0) return 0;

  const uploadAudits = auditLogs
    .filter((l) => l.action === 'wholesale_order_upload_payment_proof')
    .map((l) => {
      const base = parseAuditChanges(l.changes);
      if (!base) return null;
      const fileCountRaw = base.file_count ?? base.files ?? 1;
      const file_count = Number.isFinite(Number(fileCountRaw)) ? Number(fileCountRaw) : 1;
      const amountRaw = base.amount;
      const amountNum = typeof amountRaw === 'number' ? amountRaw : Number(amountRaw);
      const amount = Number.isFinite(amountNum) ? amountNum : undefined;
      return { created_at: l.created_at, file_count, amount };
    })
    .filter(Boolean) as { created_at: string; file_count: number; amount?: number }[];

  if (uploadAudits.length === 0) return 0;

  const docsSorted = [...proofDocs].sort(
    (a, b) => new Date(a.created_at).getTime() - new Date(b.created_at).getTime(),
  );
  const auditsSorted = [...uploadAudits].sort(
    (a, b) => new Date(a.created_at).getTime() - new Date(b.created_at).getTime(),
  );

  let docIndex = 0;
  let totalProofAmount = 0;
  for (const audit of auditsSorted) {
    let remaining = audit.file_count || 1;
    const perFileAmount =
      audit.amount != null && remaining > 0 ? audit.amount / remaining : undefined;
    while (remaining > 0 && docIndex < docsSorted.length) {
      if (perFileAmount != null) {
        totalProofAmount += perFileAmount;
      }
      docIndex += 1;
      remaining -= 1;
    }
  }
  return totalProofAmount;
}

export function isPaymentConfirmationStepComplete(
  order: WholesaleOrder,
  ctx: WholesaleOrderWorkflowContext,
): boolean {
  // Order complete date (payment confirmed) always completes this step.
  if (order.payment_confirmed_at) return true;

  const orderTotal = ctx.orderTotal ?? wholesaleOrderGrandTotal(order);
  const totalProofAmount = resolveTotalProofAmount(order, ctx);

  if (totalProofAmount != null) {
    const pending = Math.max(0, orderTotal - totalProofAmount);
    return pending < 0.01;
  }
  if (hasPaymentProofDocument(order)) {
    return false;
  }
  return false;
}

/** True when payment is fully received (confirmed date or proof totals cover order total). */
export function isWholesaleOrderPaymentSettled(
  order: WholesaleOrder,
  ctx: WholesaleOrderWorkflowContext = {},
): boolean {
  if (order.payment_confirmed_at) return true;

  const orderTotal = ctx.orderTotal ?? wholesaleOrderGrandTotal(order);
  const totalProofAmount = resolveTotalProofAmount(order, ctx);
  if (totalProofAmount != null) {
    return Math.max(0, orderTotal - totalProofAmount) < 0.01;
  }
  return false;
}

function orderLinesHaveAssignedStore(order: WholesaleOrder): boolean {
  return (
    (order.items?.length ?? 0) > 0 &&
    order.items!.every((it) => it.assigned_store_id != null)
  );
}

/** Works on list payloads that omit shipment line items or nested wholesale_order_item qty. */
function isAssignmentComplete(order: WholesaleOrder, auditLogs?: AuditLog[]): boolean {
  if (allOrderLinesFullyAssigned(order)) return true;
  if (auditLogs?.some((l) => l.action === 'wholesale_order_complete_assignment')) {
    return true;
  }
  const shipments = order.shipments ?? [];
  if (shipments.length === 0) return false;
  const shipmentLinesLoaded = shipments.some((s) => (s.items?.length ?? 0) > 0);
  if (!shipmentLinesLoaded) {
    return orderLinesHaveAssignedStore(order);
  }
  // List payloads may omit nested shipment line qty while detail loads full rows.
  // Once stores are assigned and shipments have started or finished, treat allocation as done.
  if (orderLinesHaveAssignedStore(order)) {
    if (shipments.some((s) => !shipmentStatusAllowsAssignmentChange(s.status))) {
      return true;
    }
  }
  return false;
}

function isOrderConfirmationComplete(order: WholesaleOrder, auditLogs?: AuditLog[]): boolean {
  if (order.status === 'pending_approval' || order.status === 'rejected' || order.status === 'deleted') {
    return false;
  }
  return isAssignmentComplete(order, auditLogs);
}

/** Shared step list for the detail stepper and list/detail status chips. */
export function computeWholesaleOrderProcessSteps(
  order: WholesaleOrder,
  ctx: WholesaleOrderWorkflowContext = {},
): WholesaleOrderProcessStep[] {
  const { allShipmentsStarted, allShipmentsCompleted } = shipmentMetrics(order);
  const hasInvoice = orderHasInvoiceDocument(order);

  const stepCreated = true;
  const stepOrderConfirmation = isOrderConfirmationComplete(order, ctx.auditLogs);
  const stepStartShipment = allShipmentsStarted;
  const stepFinishShipment = allShipmentsCompleted;
  const stepSendInvoiceEmail = isWholesaleOrderInvoiceEmailDone(order, ctx.auditLogs);
  const stepPaymentConfirmation = isPaymentConfirmationStepComplete(order, ctx);
  const stepComplete =
    stepOrderConfirmation &&
    stepFinishShipment &&
    stepSendInvoiceEmail &&
    stepPaymentConfirmation;

  const steps: WholesaleOrderProcessStep[] = [
    { labelKey: 'stepCreate', done: stepCreated },
    { labelKey: 'stepOrderConfirmation', done: stepOrderConfirmation },
    { labelKey: 'stepStartShipment', done: stepStartShipment },
    { labelKey: 'stepFinishShipment', done: stepFinishShipment },
  ];

  if (hasInvoice) {
    steps.push({ labelKey: 'stepSendInvoiceEmail', done: stepSendInvoiceEmail });
  }

  steps.push(
    { labelKey: 'stepPaymentConfirmation', done: stepPaymentConfirmation },
    { labelKey: 'stepComplete', done: stepComplete },
  );

  return steps;
}

export function getCurrentWholesaleOrderProcessStepKey(
  steps: WholesaleOrderProcessStep[],
): WholesaleOrderProcessStepKey | null {
  return steps.find((s) => !s.done)?.labelKey ?? null;
}

export function isWholesaleOrderPaymentConfirmationPhase(
  order: WholesaleOrder,
  ctx: WholesaleOrderWorkflowContext = {},
): boolean {
  return (
    getCurrentWholesaleOrderProcessStepKey(computeWholesaleOrderProcessSteps(order, ctx)) ===
    'stepPaymentConfirmation'
  );
}

/** @deprecated Use isWholesaleOrderPaymentConfirmationPhase */
export function isWholesaleOrderAwaitingPaymentPhase(
  order: WholesaleOrder,
  ctx: WholesaleOrderWorkflowContext = {},
): boolean {
  return isWholesaleOrderPaymentConfirmationPhase(order, ctx);
}

/** Status chip label — operational state (what is pending now), not the stepper milestone name. */
export function wholesaleOrderStatusLabel(
  order: WholesaleOrder,
  t: TFunction,
  ctx: WholesaleOrderWorkflowContext = {},
): string {
  if (order.status === 'rejected') return t('wholesaleOrdersPage:statusRejected');
  if (order.status === 'deleted') return t('wholesaleOrdersPage:statusDeleted');

  if (isWholesaleOrderPaymentSettled(order, ctx)) {
    return t('wholesaleOrderDetail:statusCompleted');
  }
  if (hasPaymentProofDocument(order)) {
    return t('wholesaleOrderDetail:orderStatusPendingPaymentConfirmation');
  }

  const steps = computeWholesaleOrderProcessSteps(order, ctx);
  if (steps.every((s) => s.done)) return t('wholesaleOrderDetail:statusCompleted');

  const current = getCurrentWholesaleOrderProcessStepKey(steps);
  switch (current) {
    case 'stepOrderConfirmation':
      return t('wholesaleOrderDetail:orderStatusPendingOrderConfirmation');
    case 'stepStartShipment':
      return t('wholesaleOrderDetail:orderStatusPendingPacking');
    case 'stepFinishShipment':
      return t('wholesaleOrderDetail:orderStatusInTransit');
    case 'stepSendInvoiceEmail':
      return t('wholesaleOrderDetail:orderStatusPendingInvoiceEmail');
    case 'stepPaymentConfirmation':
      return t('wholesaleOrderDetail:orderStatusPendingPaymentConfirmation');
    case 'stepComplete':
      return t('wholesaleOrderDetail:statusCompleted');
    default:
      return String(order.status).replace(/_/g, ' ');
  }
}

/** Status chip color aligned with the first incomplete process step. */
export function wholesaleOrderStatusColor(
  order: WholesaleOrder,
  ctx: WholesaleOrderWorkflowContext = {},
): WholesaleOrderStatusChipColor {
  if (order.status === 'rejected') return 'error';
  if (order.status === 'deleted') return 'default';

  if (isWholesaleOrderPaymentSettled(order, ctx)) return 'success';
  if (hasPaymentProofDocument(order)) return 'secondary';

  const steps = computeWholesaleOrderProcessSteps(order, ctx);
  if (steps.every((s) => s.done)) return 'success';

  const current = getCurrentWholesaleOrderProcessStepKey(steps);
  switch (current) {
    case 'stepOrderConfirmation':
      return 'primary';
    case 'stepStartShipment':
      return 'warning';
    case 'stepFinishShipment':
      return 'info';
    case 'stepSendInvoiceEmail':
      return 'warning';
    case 'stepPaymentConfirmation':
      return 'secondary';
    default:
      return 'default';
  }
}

export function getWholesaleOrderProcessStepCompletedAt(
  order: WholesaleOrder,
  labelKey: WholesaleOrderProcessStepKey,
  opts: {
    auditLogs?: AuditLog[];
    assignmentCompletedAt?: string | null;
    stepFinishShipment?: boolean;
  } = {},
): string | null {
  const steps = computeWholesaleOrderProcessSteps(order, {
    auditLogs: opts.auditLogs,
  });
  const step = steps.find((s) => s.labelKey === labelKey);
  if (!step?.done) return null;

  const auditLogs = opts.auditLogs ?? [];
  const emailAudits = getWholesaleOrderEmailAudits(auditLogs);

  switch (labelKey) {
    case 'stepCreate':
      return order.created_at ?? null;
    case 'stepOrderConfirmation':
      return opts.assignmentCompletedAt ?? order.reviewed_at ?? null;
    case 'stepStartShipment':
      if (order.shipments?.length) {
        return order.shipments.reduce<string | null>((a, s) => {
          const at = s.created_at;
          if (!at) return a;
          return !a || at < a ? at : a;
        }, null);
      }
      return null;
    case 'stepFinishShipment':
      if (opts.stepFinishShipment && order.shipments?.length) {
        return order.shipments
          .filter((s) => s.status === 'completed')
          .reduce<string | null>((a, s) => {
            const at = s.updated_at ?? s.delivery_date ?? s.created_at;
            if (!at) return a;
            return !a || at > a ? at : a;
          }, null);
      }
      return null;
    case 'stepSendInvoiceEmail': {
      const changes = emailAudits.invoice
        ? parseWholesaleOrderEmailAuditBase(emailAudits.invoice.changes)
        : null;
      if (changes && isWholesaleOrderEmailSkippedAudit(changes)) {
        return (
          (typeof changes.skipped_at === 'string' && changes.skipped_at.trim()
            ? changes.skipped_at.trim()
            : null) ?? emailAudits.invoice?.created_at ?? null
        );
      }
      if (changes && isWholesaleOrderEmailSentAudit(changes)) {
        return (
          (typeof changes.sent_at === 'string' && changes.sent_at.trim()
            ? changes.sent_at.trim()
            : null) ?? emailAudits.invoice?.created_at ?? null
        );
      }
      return null;
    }
    case 'stepPaymentConfirmation':
    case 'stepComplete':
      return order.payment_confirmed_at ?? null;
    default:
      return null;
  }
}

File: management_mobile/lib/utils/wholesale_order_workflow.dart
import 'dart:convert';

import '../models/admin.dart';
import '../models/shipment.dart';
import '../models/wholesale_order.dart';
import 'status_chip_style.dart';
import 'wholesale_order_assignment.dart';

enum WholesaleProcessStepKey {
  stepCreate,
  stepOrderConfirmation,
  stepStartShipment,
  stepFinishShipment,
  stepSendInvoiceEmail,
  stepPaymentConfirmation,
  stepComplete,
}

class WholesaleProcessStep {
  const WholesaleProcessStep({required this.key, required this.done});
  final WholesaleProcessStepKey key;
  final bool done;
}

class WholesaleWorkflowContext {
  const WholesaleWorkflowContext({this.auditLogs = const [], this.totalProofAmount});
  final List<AuditLogEntry> auditLogs;
  final double? totalProofAmount;
}

bool shipmentHasDeliveryNoteStarted(Shipment shipment) {
  return shipment.status == 'packed' ||
      shipment.status == 'shipped' ||
      shipment.status == 'completed' ||
      (shipment.deliveryNotePdfUrl?.trim().isNotEmpty ?? false);
}

bool orderHasInvoiceDocument(WholesaleOrder order) {
  return order.documents.any((d) => d.type == 'invoice');
}

bool hasPaymentProofDocument(WholesaleOrder order) {
  if (order.paymentProofUrl?.trim().isNotEmpty ?? false) return true;
  return order.documents.any((d) => d.type == 'payment_proof');
}

double wholesaleOrderGrandTotal(WholesaleOrder order) {
  final itemsTotal = order.totalNet ?? order.itemsTotal;
  return itemsTotal + (order.shippingFee ?? 0);
}

Map<String, dynamic>? _parseAuditChanges(String? raw) {
  if (raw == null || raw.isEmpty) return null;
  try {
    final parsed = jsonDecode(raw);
    if (parsed is! Map<String, dynamic>) return null;
    final nested = parsed['changes'];
    if (nested is Map<String, dynamic>) return nested;
    return parsed;
  } catch (_) {
    return null;
  }
}

bool _invoiceEmailDone(WholesaleOrder order, List<AuditLogEntry> auditLogs) {
  if (!orderHasInvoiceDocument(order)) return true;
  for (final log in auditLogs) {
    if (!log.action.contains('email') || log.changes == null) continue;
    final changes = _parseAuditChanges(log.changes);
    if (changes == null) continue;
    final emailType = changes['email_type']?.toString();
    if (emailType != 'invoice') continue;
    if (changes['skipped_at'] != null || changes['sent_at'] != null || changes['done'] == true) {
      return true;
    }
  }
  if (order.workflowInvoiceEmailDone) return true;
  if (order.invoiceSentAt?.trim().isNotEmpty ?? false) return true;
  if (order.paymentConfirmedAt?.trim().isNotEmpty ?? false) return true;
  return false;
}

double computeTotalProofAmountFromAudits(WholesaleOrder order, List<AuditLogEntry> auditLogs) {
  final proofDocs = order.documents.where((d) => d.type == 'payment_proof').toList();
  if (proofDocs.isEmpty) return 0;

  final uploadAudits = auditLogs
      .where((l) => l.action == 'wholesale_order_upload_payment_proof')
      .map((l) {
        final base = _parseAuditChanges(l.changes);
        if (base == null) return null;
        final fileCountRaw = base['file_count'] ?? base['files'] ?? 1;
        final fileCount = double.tryParse('$fileCountRaw') ?? 1;
        final amountNum = double.tryParse('${base['amount']}');
        return (createdAt: l.createdAt, fileCount: fileCount, amount: amountNum);
      })
      .whereType<({String? createdAt, double fileCount, double? amount})>()
      .toList();

  if (uploadAudits.isEmpty) return order.workflowPaymentProofTotal ?? 0;

  proofDocs.sort((a, b) => (a.createdAt ?? '').compareTo(b.createdAt ?? ''));
  uploadAudits.sort((a, b) => (a.createdAt ?? '').compareTo(b.createdAt ?? ''));

  var docIndex = 0;
  var total = 0.0;
  for (final audit in uploadAudits) {
    var remaining = audit.fileCount <= 0 ? 1.0 : audit.fileCount;
    final perFile = audit.amount != null && remaining > 0 ? audit.amount! / remaining : null;
    while (remaining > 0 && docIndex < proofDocs.length) {
      if (perFile != null) total += perFile;
      docIndex += 1;
      remaining -= 1;
    }
  }
  return total;
}

WholesaleWorkflowContext buildWholesaleWorkflowContext(
  WholesaleOrder order,
  List<AuditLogEntry> auditLogs,
) {
  final totalProof = auditLogs.isNotEmpty
      ? computeTotalProofAmountFromAudits(order, auditLogs)
      : order.workflowPaymentProofTotal;
  return WholesaleWorkflowContext(
    auditLogs: auditLogs,
    totalProofAmount: totalProof,
  );
}

bool isPaymentConfirmationStepComplete(WholesaleOrder order, WholesaleWorkflowContext ctx) {
  // Order complete date (payment confirmed) always completes this step.
  if (order.paymentConfirmedAt != null && order.paymentConfirmedAt!.trim().isNotEmpty) {
    return true;
  }

  final orderTotal = wholesaleOrderGrandTotal(order);
  final totalProof = ctx.totalProofAmount;
  if (totalProof != null) {
    return orderTotal - totalProof < 0.01;
  }
  if (hasPaymentProofDocument(order)) return false;
  return false;
}

bool isWholesaleOrderPaymentSettled(WholesaleOrder order, WholesaleWorkflowContext ctx) {
  if (order.paymentConfirmedAt != null && order.paymentConfirmedAt!.trim().isNotEmpty) {
    return true;
  }
  final orderTotal = wholesaleOrderGrandTotal(order);
  final totalProof = ctx.totalProofAmount ?? order.workflowPaymentProofTotal;
  if (totalProof != null) {
    return orderTotal - totalProof < 0.01;
  }
  return false;
}

List<String> _completedAssignmentActions(List<AuditLogEntry> auditLogs) {
  return auditLogs
      .where((l) => l.action == 'wholesale_order_complete_assignment')
      .map((l) => l.action)
      .toList();
}

List<WholesaleProcessStep> computeWholesaleOrderProcessSteps(
  WholesaleOrder order,
  WholesaleWorkflowContext ctx,
) {
  final shipments = order.shipments;
  final hasShipments = shipments.isNotEmpty;
  final allStarted = hasShipments && shipments.every(shipmentHasDeliveryNoteStarted);
  final allCompleted = hasShipments && shipments.every((s) => s.status == 'completed');
  final hasInvoice = orderHasInvoiceDocument(order);
  final assignmentActions = _completedAssignmentActions(ctx.auditLogs);

  final stepOrderConfirmation = order.status != 'pending_approval' &&
      order.status != 'rejected' &&
      order.status != 'deleted' &&
      isAssignmentComplete(order, completedAssignmentActions: assignmentActions);
  final stepSendInvoiceEmail = _invoiceEmailDone(order, ctx.auditLogs);
  final stepPaymentConfirmation = isPaymentConfirmationStepComplete(order, ctx);

  final steps = <WholesaleProcessStep>[
    const WholesaleProcessStep(key: WholesaleProcessStepKey.stepCreate, done: true),
    WholesaleProcessStep(key: WholesaleProcessStepKey.stepOrderConfirmation, done: stepOrderConfirmation),
    WholesaleProcessStep(key: WholesaleProcessStepKey.stepStartShipment, done: allStarted),
    WholesaleProcessStep(key: WholesaleProcessStepKey.stepFinishShipment, done: allCompleted),
  ];

  if (hasInvoice) {
    steps.add(WholesaleProcessStep(key: WholesaleProcessStepKey.stepSendInvoiceEmail, done: stepSendInvoiceEmail));
  }

  steps.addAll([
    WholesaleProcessStep(key: WholesaleProcessStepKey.stepPaymentConfirmation, done: stepPaymentConfirmation),
    WholesaleProcessStep(
      key: WholesaleProcessStepKey.stepComplete,
      done: stepOrderConfirmation && allCompleted && stepSendInvoiceEmail && stepPaymentConfirmation,
    ),
  ]);

  return steps;
}

WholesaleProcessStepKey? currentWholesaleProcessStepKey(List<WholesaleProcessStep> steps) {
  for (final step in steps) {
    if (!step.done) return step.key;
  }
  return null;
}

bool isWholesaleOrderCompleted(WholesaleOrder order) {
  if (order.isCompleted) return true;
  if (order.shipments.isEmpty) return false;
  final allShipmentsCompleted = order.shipments.every((s) => s.status == 'completed');
  return allShipmentsCompleted && (order.paymentConfirmedAt?.trim().isNotEmpty ?? false);
}

bool isWholesaleOrderPaymentConfirmationPhase(
  WholesaleOrder order, [
  WholesaleWorkflowContext? ctx,
]) {
  final workflowCtx = ctx ?? buildWholesaleWorkflowContext(order, const []);
  final steps = computeWholesaleOrderProcessSteps(order, workflowCtx);
  return currentWholesaleProcessStepKey(steps) == WholesaleProcessStepKey.stepPaymentConfirmation;
}

/// API status query — derived filters fetch `approved` then filter client-side (portal parity).
String? wholesaleOrderStatusFilterForApi(String? statusFilter) {
  if (statusFilter == null || statusFilter.isEmpty) return null;
  if (statusFilter == 'awaiting_payment' || statusFilter == 'completed') {
    return 'approved';
  }
  return statusFilter;
}

List<WholesaleOrder> filterWholesaleOrdersByStatus(List<WholesaleOrder> orders, String? statusFilter) {
  if (statusFilter == null || statusFilter.isEmpty) return orders;
  switch (statusFilter) {
    case 'awaiting_payment':
      return orders.where(isWholesaleOrderPaymentConfirmationPhase).toList();
    case 'completed':
      return orders.where(isWholesaleOrderCompleted).toList();
    default:
      return orders.where((o) => o.status == statusFilter).toList();
  }
}

const wholesaleOrderStatusFilterOptions = <({String? value, String label})>[
  (value: null, label: 'All'),
  (value: 'pending_approval', label: 'Pending approval'),
  (value: 'assign_shipment', label: 'Assign shipment'),
  (value: 'awaiting_payment', label: 'Awaiting payment'),
  (value: 'completed', label: 'Completed'),
  (value: 'rejected', label: 'Rejected'),
  (value: 'deleted', label: 'Deleted'),
];

bool wholesaleOrderAllowsStoreAssignment(
  WholesaleOrder order,
  WholesaleProcessStepKey? currentStep, {
  Iterable<String>? completedAssignmentActions,
}) {
  if (order.status != 'approved' && order.status != 'assign_shipment') return false;
  if (currentStep != WholesaleProcessStepKey.stepOrderConfirmation) return false;
  if (isAssignmentComplete(order, completedAssignmentActions: completedAssignmentActions)) {
    return false;
  }
  return orderAllowsAssignmentChange(order);
}

bool wholesaleOrderShowsAssignmentPanel({
  required WholesaleOrder order,
  required WholesaleProcessStepKey? currentStep,
  required bool allocationConfirmed,
  required List<StagedStoreAssignment> staged,
  Iterable<String>? completedAssignmentActions,
}) {
  if (!wholesaleOrderCanAssign(order)) return false;
  if (order.status == 'pending_approval') {
    return !allocationConfirmed || !allOrderLinesFullyStaged(order, staged);
  }
  // Match portal: once allocation is confirmed show summary; re-open editor when user changes assignment.
  if (!allocationConfirmed) return true;
  return false;
}

String wholesaleProcessStepLabel(WholesaleProcessStepKey key) {
  switch (key) {
    case WholesaleProcessStepKey.stepCreate:
      return 'Created';
    case WholesaleProcessStepKey.stepOrderConfirmation:
      return 'Order confirmation';
    case WholesaleProcessStepKey.stepStartShipment:
      return 'Start shipment';
    case WholesaleProcessStepKey.stepFinishShipment:
      return 'Finish shipment';
    case WholesaleProcessStepKey.stepSendInvoiceEmail:
      return 'Invoice email';
    case WholesaleProcessStepKey.stepPaymentConfirmation:
      return 'Payment confirmation';
    case WholesaleProcessStepKey.stepComplete:
      return 'Complete';
  }
}

String wholesaleOrderWorkflowStatusLabel(WholesaleOrder order, WholesaleWorkflowContext ctx) {
  if (order.status == 'rejected') return 'Rejected';
  if (order.status == 'deleted') return 'Deleted';

  if (isWholesaleOrderPaymentSettled(order, ctx)) return 'Completed';
  if (hasPaymentProofDocument(order)) return 'Pending payment confirmation';

  final steps = computeWholesaleOrderProcessSteps(order, ctx);
  if (steps.every((s) => s.done)) return 'Completed';

  switch (currentWholesaleProcessStepKey(steps)) {
    case WholesaleProcessStepKey.stepOrderConfirmation:
      return 'Pending order confirmation';
    case WholesaleProcessStepKey.stepStartShipment:
      return 'Pending packing';
    case WholesaleProcessStepKey.stepFinishShipment:
      return 'In transit';
    case WholesaleProcessStepKey.stepSendInvoiceEmail:
      return 'Pending invoice email';
    case WholesaleProcessStepKey.stepPaymentConfirmation:
      return 'Pending payment confirmation';
    case WholesaleProcessStepKey.stepComplete:
      return 'Completed';
    default:
      return wholesaleStatusLabel(order.status);
  }
}

StatusChipColor wholesaleOrderStatusColor(WholesaleOrder order, WholesaleWorkflowContext ctx) {
  if (order.status == 'rejected') return StatusChipColor.error;
  if (order.status == 'deleted') return StatusChipColor.defaultColor;

  if (isWholesaleOrderPaymentSettled(order, ctx)) return StatusChipColor.success;
  if (hasPaymentProofDocument(order)) return StatusChipColor.secondary;

  final steps = computeWholesaleOrderProcessSteps(order, ctx);
  if (steps.every((s) => s.done)) return StatusChipColor.success;

  switch (currentWholesaleProcessStepKey(steps)) {
    case WholesaleProcessStepKey.stepOrderConfirmation:
      return StatusChipColor.primary;
    case WholesaleProcessStepKey.stepStartShipment:
      return StatusChipColor.warning;
    case WholesaleProcessStepKey.stepFinishShipment:
      return StatusChipColor.info;
    case WholesaleProcessStepKey.stepSendInvoiceEmail:
      return StatusChipColor.warning;
    case WholesaleProcessStepKey.stepPaymentConfirmation:
      return StatusChipColor.secondary;
    default:
      return StatusChipColor.defaultColor;
  }
}

String wholesaleStatusLabel(String status) {
  switch (status) {
    case 'pending_approval':
      return 'Pending approval';
    case 'assign_shipment':
      return 'Assign shipment';
    case 'approved':
      return 'Approved';
    case 'rejected':
      return 'Rejected';
    case 'deleted':
      return 'Deleted';
    default:
      return status.replaceAll('_', ' ');
  }
}
